diff --git a/ext/wayland/gstwaylandsink.c b/ext/wayland/gstwaylandsink.c
index 2f116bf..01404c5 100644
--- a/ext/wayland/gstwaylandsink.c
+++ b/ext/wayland/gstwaylandsink.c
@@ -342,6 +342,19 @@ gst_wayland_sink_set_display_from_context (GstWaylandSink * self,
   }
 }
 
+static void
+on_pointer_connected (GstWlDisplay *display, GstWaylandSink wl_sink)
+{
+  gst_wl_display_enable_pointer_handle (display);
+  /*removing pointer on disconnect handled internally to wldisplay*/
+}
+
+static void
+on_pointer_move (GstWlDisplay *display, int handle,  GstWaylandSink wl_sink)
+{
+  gst_wl_window_move(wl_sink.window, gst_wl_display_get_seat(display), handle);
+}
+
 static gboolean
 gst_wayland_sink_find_display (GstWaylandSink * self)
 {
@@ -385,6 +398,20 @@ gst_wayland_sink_find_display (GstWaylandSink * self)
               ("Failed to create GstWlDisplay: '%s'", error->message));
           g_error_free (error);
           ret = FALSE;
+        } else {
+          /* for locally created display and window is not fullscreen,
+           * handle the pointer events as required */
+          if (!self->fullscreen) {
+            g_signal_connect(self->display, "pointer-connected",
+              G_CALLBACK(on_pointer_connected), self);
+            g_signal_connect(self->display, "pointer-move",
+              G_CALLBACK(on_pointer_move), self);
+
+            /* handle initial state */
+            if (gst_wl_display_is_mouse_connected(self->display))
+              gst_wl_display_enable_pointer_handle (self->display);
+
+          }
         }
       }
     }
diff --git a/gst-libs/gst/wayland/gstwldisplay.c b/gst-libs/gst/wayland/gstwldisplay.c
index 1f83cb2..35a9fa2 100644
--- a/gst-libs/gst/wayland/gstwldisplay.c
+++ b/gst-libs/gst/wayland/gstwldisplay.c
@@ -28,7 +28,7 @@
 #include "linux-dmabuf-unstable-v1-client-protocol.h"
 #include "viewporter-client-protocol.h"
 #include "xdg-shell-client-protocol.h"
-
+#include "wayland-cursor.h"
 #include <errno.h>
 
 #define GST_CAT_DEFAULT gst_wl_display_debug
@@ -52,6 +52,7 @@ typedef struct _GstWlDisplayPrivate
   struct zwp_linux_dmabuf_v1 *dmabuf;
   GArray *shm_formats;
   GArray *dmabuf_formats;
+  struct wl_seat *seat;
 
   /* private */
   gboolean own_display;
@@ -61,6 +62,11 @@ typedef struct _GstWlDisplayPrivate
   GMutex buffers_mutex;
   GHashTable *buffers;
   gboolean shutting_down;
+  struct wl_pointer *pointer;
+  struct wl_surface *cursor_surface;
+  struct wl_cursor_theme *cursor_theme;
+  struct wl_cursor *default_cursor;
+  enum wl_seat_capability caps;
 } GstWlDisplayPrivate;
 
 G_DEFINE_TYPE_WITH_CODE (GstWlDisplay, gst_wl_display, G_TYPE_OBJECT,
@@ -69,6 +75,15 @@ G_DEFINE_TYPE_WITH_CODE (GstWlDisplay, gst_wl_display, G_TYPE_OBJECT,
         "wldisplay", 0, "wldisplay library");
     );
 
+enum
+{
+  EVENT_POINTER_CONNECTION,
+  EVENT_POINTER_MOVE,
+  LAST_SIGNAL
+};
+
+static guint gst_wl_display_signals[LAST_SIGNAL] = { 0 };
+
 static void gst_wl_display_finalize (GObject * gobject);
 
 static void
@@ -76,6 +91,14 @@ gst_wl_display_class_init (GstWlDisplayClass * klass)
 {
   GObjectClass *gobject_class = G_OBJECT_CLASS (klass);
   gobject_class->finalize = gst_wl_display_finalize;
+
+  gst_wl_display_signals[EVENT_POINTER_CONNECTION] =
+    g_signal_new ("pointer-connected", G_TYPE_FROM_CLASS (klass),
+    G_SIGNAL_RUN_LAST, 0, NULL, NULL, NULL, G_TYPE_NONE, 0);
+
+  gst_wl_display_signals[EVENT_POINTER_MOVE] =
+    g_signal_new ("pointer-move", G_TYPE_FROM_CLASS (klass),
+    G_SIGNAL_RUN_LAST, 0, NULL, NULL, NULL, G_TYPE_NONE, 1, G_TYPE_INT);
 }
 
 static void
@@ -249,6 +272,101 @@ static const struct xdg_wm_base_listener xdg_wm_base_listener = {
   handle_xdg_wm_base_ping
 };
 
+static void
+pointer_handle_enter(void *data, struct wl_pointer *pointer,
+         uint32_t serial, struct wl_surface *surface,
+         wl_fixed_t sx, wl_fixed_t sy)
+{
+  GstWlDisplay *display = data;
+  GstWlDisplayPrivate *priv = gst_wl_display_get_instance_private (display);
+  struct wl_buffer *buffer;
+  struct wl_cursor_image *image;
+
+  if (!priv->default_cursor) {
+    GST_WARNING_OBJECT (display, "cursor image not loaded, pointer not set");
+    return;
+  }
+
+  image = priv->default_cursor->images[0];
+  buffer = wl_cursor_image_get_buffer(image);
+  if (!buffer)
+    return;
+
+  if (!priv->cursor_surface) {
+    priv->cursor_surface = wl_compositor_create_surface(priv->compositor);
+  }
+  wl_pointer_set_cursor(priv->pointer, serial,
+                        priv->cursor_surface,
+                        image->hotspot_x,
+                        image->hotspot_y);
+
+  wl_surface_attach(priv->cursor_surface, buffer, 0, 0);
+  wl_surface_damage(priv->cursor_surface, 0, 0,
+                    image->width, image->height);
+  wl_surface_commit(priv->cursor_surface);
+}
+
+static void
+pointer_handle_leave(void *data, struct wl_pointer *pointer,
+         uint32_t serial, struct wl_surface *surface)
+{
+}
+
+static void
+pointer_handle_motion(void *data, struct wl_pointer *pointer,
+          uint32_t time, wl_fixed_t sx, wl_fixed_t sy)
+{
+}
+
+static void
+pointer_handle_button(void *data, struct wl_pointer *wl_pointer,
+          uint32_t serial, uint32_t time, uint32_t button,
+          uint32_t state)
+{
+  GstWlDisplay *display = data;
+
+  if (button == /*BTN_LEFT*/0x110 && state == WL_POINTER_BUTTON_STATE_PRESSED) {
+    g_signal_emit(display, gst_wl_display_signals[EVENT_POINTER_MOVE], 0, serial);
+  }
+}
+
+static void
+pointer_handle_axis(void *data, struct wl_pointer *wl_pointer,
+                    uint32_t time, uint32_t axis, wl_fixed_t value)
+{
+}
+
+static const struct wl_pointer_listener pointer_listener = {
+  pointer_handle_enter,
+  pointer_handle_leave,
+  pointer_handle_motion,
+  pointer_handle_button,
+  pointer_handle_axis,
+};
+
+static void
+seat_handle_capabilities(void *data, struct wl_seat *seat,
+                         enum wl_seat_capability caps)
+{
+  GstWlDisplay *self = data;
+  GstWlDisplayPrivate *priv = gst_wl_display_get_instance_private (self);
+
+  priv->caps = caps;
+  if ((caps & WL_SEAT_CAPABILITY_POINTER) && !priv->pointer) {
+    /*wlsink to decide on whether to enable pointer*/
+    g_signal_emit(self, gst_wl_display_signals[EVENT_POINTER_CONNECTION], 0);
+  } else if (!(caps & WL_SEAT_CAPABILITY_POINTER) && priv->pointer) {
+    if (priv->pointer)
+      wl_pointer_destroy (priv->pointer);
+    priv->pointer = NULL;
+  }
+}
+
+
+static const struct wl_seat_listener seat_listener = {
+  seat_handle_capabilities
+};
+
 static void
 registry_handle_global (void *data, struct wl_registry *registry,
     uint32_t id, const char *interface, uint32_t version)
@@ -272,6 +390,12 @@ registry_handle_global (void *data, struct wl_registry *registry,
   } else if (g_strcmp0 (interface, "wl_shm") == 0) {
     priv->shm = wl_registry_bind (registry, id, &wl_shm_interface, 1);
     wl_shm_add_listener (priv->shm, &shm_listener, self);
+
+    /*load cursor theme*/
+    priv->cursor_theme = wl_cursor_theme_load(NULL, 32, priv->shm);
+    if (priv->cursor_theme)
+      priv->default_cursor = wl_cursor_theme_get_cursor(priv->cursor_theme, "left_ptr");
+
   } else if (g_strcmp0 (interface, "wp_viewporter") == 0) {
     priv->viewporter =
         wl_registry_bind (registry, id, &wp_viewporter_interface, 1);
@@ -279,6 +403,9 @@ registry_handle_global (void *data, struct wl_registry *registry,
     priv->dmabuf =
         wl_registry_bind (registry, id, &zwp_linux_dmabuf_v1_interface, 1);
     zwp_linux_dmabuf_v1_add_listener (priv->dmabuf, &dmabuf_listener, self);
+  } else if (strcmp(interface, "wl_seat") == 0) {
+    priv->seat = wl_registry_bind(registry, id, &wl_seat_interface, 1);
+    wl_seat_add_listener(priv->seat, &seat_listener, self);
   }
 }
 
@@ -565,3 +692,25 @@ gst_wl_display_has_own_display (GstWlDisplay * self)
 
   return priv->own_display;
 }
+
+struct wl_seat *
+gst_wl_display_get_seat(GstWlDisplay * self)
+{
+  GstWlDisplayPrivate *priv = gst_wl_display_get_instance_private (self);
+
+  return priv->seat;
+}
+
+void gst_wl_display_enable_pointer_handle (GstWlDisplay * self)
+{
+  GST_TRACE_OBJECT (self, "enable mouse pointer handling");
+  GstWlDisplayPrivate *priv = gst_wl_display_get_instance_private (self);
+  priv->pointer = wl_seat_get_pointer (priv->seat);
+  wl_pointer_add_listener(priv->pointer, &pointer_listener, self);
+}
+
+gboolean gst_wl_display_is_mouse_connected (GstWlDisplay * self)
+{
+  GstWlDisplayPrivate *priv = gst_wl_display_get_instance_private (self);
+  return (priv->caps & WL_SEAT_CAPABILITY_POINTER) ? 1 : 0;
+}
diff --git a/gst-libs/gst/wayland/gstwldisplay.h b/gst-libs/gst/wayland/gstwldisplay.h
index eb07e4f..870bf4a 100644
--- a/gst-libs/gst/wayland/gstwldisplay.h
+++ b/gst-libs/gst/wayland/gstwldisplay.h
@@ -97,4 +97,13 @@ struct zwp_linux_dmabuf_v1 *gst_wl_display_get_dmabuf_v1 (GstWlDisplay * self);
 GST_WL_API
 gboolean gst_wl_display_has_own_display (GstWlDisplay * self);
 
+GST_WL_API
+struct wl_seat * gst_wl_display_get_seat(GstWlDisplay * self);
+
+GST_WL_API
+void gst_wl_display_enable_pointer_handle (GstWlDisplay * self);
+
+GST_WL_API
+gboolean gst_wl_display_is_mouse_connected (GstWlDisplay * self);
+
 G_END_DECLS
diff --git a/gst-libs/gst/wayland/gstwlwindow.c b/gst-libs/gst/wayland/gstwlwindow.c
index 0023d4e..235f7b0 100644
--- a/gst-libs/gst/wayland/gstwlwindow.c
+++ b/gst-libs/gst/wayland/gstwlwindow.c
@@ -687,3 +687,12 @@ gst_wl_window_set_rotate_method (GstWlWindow * self,
 
   gst_wl_window_update_geometry (self);
 }
+
+void
+gst_wl_window_move (GstWlWindow * self, struct wl_seat *seat, uint32_t serial)
+{
+  GstWlWindowPrivate *priv = gst_wl_window_get_instance_private (self);
+
+  if (priv->xdg_toplevel)
+    xdg_toplevel_move(priv->xdg_toplevel, seat, serial);
+}
diff --git a/gst-libs/gst/wayland/gstwlwindow.h b/gst-libs/gst/wayland/gstwlwindow.h
index 06c4001..d221940 100644
--- a/gst-libs/gst/wayland/gstwlwindow.h
+++ b/gst-libs/gst/wayland/gstwlwindow.h
@@ -72,4 +72,7 @@ GST_WL_API
 void gst_wl_window_set_rotate_method (GstWlWindow               *self,
         GstVideoOrientationMethod  rotate_method);
 
+GST_WL_API
+void gst_wl_window_move (GstWlWindow * self, struct wl_seat *seat, uint32_t serial);
+
 G_END_DECLS
diff --git a/gst-libs/gst/wayland/meson.build b/gst-libs/gst/wayland/meson.build
index 5eb4484..628d0f0 100644
--- a/gst-libs/gst/wayland/meson.build
+++ b/gst-libs/gst/wayland/meson.build
@@ -3,6 +3,8 @@ wl_client_dep = dependency('wayland-client', version: wl_req, required: get_opti
 libdrm_dep = dependency('libdrm', version: '>= 2.4.55', required: get_option('wayland'))
 wl_protocol_dep = dependency('wayland-protocols', version: wl_req, required: get_option('wayland'))
 wl_scanner = find_program('wayland-scanner', required: get_option('wayland'))
+wl_cursor_dep = dependency('wayland-cursor', version: wl_req, required: get_option('wayland'))
+
 # Also used in ext/wayland
 use_wayland = wl_protocol_dep.found() and wl_client_dep.found() and wl_scanner.found() and libdrm_dep.found()
 
@@ -74,7 +76,7 @@ if use_wayland
     darwin_versions : osxversion,
     install : true,
     dependencies : [gst_dep, gstallocators_dep, gstvideo_dep, libdrm_dep,
-                    wl_client_dep]
+                    wl_client_dep, wl_cursor_dep]
   )
 
   pkg_name = 'gstreamer-wayland-1.0'
