diff --git a/sys/kms/gstkmsallocator.c b/sys/kms/gstkmsallocator.c
index 6687f3b..74fd317 100644
--- a/sys/kms/gstkmsallocator.c
+++ b/sys/kms/gstkmsallocator.c
@@ -46,6 +46,16 @@
 #define DRM_RDWR O_RDWR
 #endif
 
+#ifndef ALIGN
+#define ALIGN(x, a)  (((x) + ((a) - 1)) & ~((typeof(x))(a) - 1))
+#endif //ALIGN
+
+#ifndef IS_ALIGNED
+#define IS_ALIGNED(x, a)  (((x) & ((typeof(x))(a) - 1)) == 0)
+#endif //IS_ALIGNED
+
+#define SIZE_4K 4096
+#define BIT_ALIGN 8
 #define GST_CAT_DEFAULT kmsallocator_debug
 GST_DEBUG_CATEGORY_STATIC (GST_CAT_DEFAULT);
 
@@ -164,6 +174,14 @@ gst_kms_allocator_memory_create (GstKMSAllocator * allocator,
   h = GST_VIDEO_INFO_HEIGHT (vinfo);
   arg.height = gst_drm_height_from_drm (fmt, h);
 
+  /*Memory is allocated from system-cust-heap and is IOMMU enabled
+   *Due to which HW expect 4k aligned plane offset in multi-planar formats
+   *size increased to compensate the offset in case of unaligned address. */
+  if ((!IS_ALIGNED(arg.width * h * (ALIGN(arg.bpp, BIT_ALIGN) >> 3), SIZE_4K)) && (num_planes > 1)) {
+    arg.height = (SIZE_4K % arg.width == 0) ? (arg.height + (SIZE_4K / arg.width))
+                   : (arg.height + (SIZE_4K / arg.width) + 1);
+  }
+
   ret = drmIoctl (allocator->priv->fd, DRM_IOCTL_MODE_CREATE_DUMB, &arg);
   if (ret)
     goto create_failed;
@@ -187,6 +205,7 @@ gst_kms_allocator_memory_create (GstKMSAllocator * allocator,
     /* Note that we cannot negotiate special padding betweem each planes,
      * hence using the display height here. */
     offs += pitch * GST_VIDEO_FORMAT_INFO_SCALE_HEIGHT (vinfo->finfo, i, h);
+    offs = ALIGN(offs, SIZE_4K);
 
     GST_DEBUG_OBJECT (allocator, "Created BO plane %i with stride %i and "
         "offset %" G_GSIZE_FORMAT, i,
