From 0e80b34a85f8f026996e71c2c554862b932ca98b Mon Sep 17 00:00:00 2001
From: Lorenzo Keller <Lorenzo.Keller@synaptics.com>
Date: Wed, 17 Jan 2024 12:38:18 +0000
Subject: [PATCH] Force gcc as HOSTCC

---
 build/module/uboot/build.sh | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/build/module/uboot/build.sh b/build/module/uboot/build.sh
index 2464654d..fc0c7887 100755
--- a/build/module/uboot/build.sh
+++ b/build/module/uboot/build.sh
@@ -7,6 +7,7 @@
 echo "u-boot build script called"
 
 build_uboot() {
+  set -x
   cd ${topdir}/boot/${CONFIG_UBOOT_SRC_PATH} || exit 1
 
   TARGET_CONFIG=merged_defconfig
@@ -32,7 +33,7 @@ build_uboot() {
   eval "$cmd2run make mrproper"
 
   cmd2run="${cmd2run} KBUILD_OUTPUT=${opt_outdir_intermediate}/output_uboot"
-  cmd2run="${cmd2run} CROSS_COMPILE=${CONFIG_TOOLCHAIN_BSP}"
+  cmd2run="${cmd2run} CROSS_COMPILE=${CONFIG_TOOLCHAIN_BSP}" # HOSTCC=\"${BUILD_CC} ${BUILD_CFLAGS} ${BUILD_LDFLAGS}\""
   cmd2run="${cmd2run}"
 
   if [ "is${CONFIG_GENX_ENABLE}" = "isy" ]; then
@@ -40,8 +41,8 @@ build_uboot() {
   fi
 
   eval "CONFIG_RDK_SYS=${CONFIG_RDK_SYS} $cmd2run make mrproper"
-  eval "CONFIG_RDK_SYS=${CONFIG_RDK_SYS} $cmd2run make ${TARGET_CONFIG} -j${CONFIG_CPU_NUMBER}"
-  eval "CONFIG_RDK_SYS=${CONFIG_RDK_SYS} $cmd2run make EXT_DTB=arch/${CONFIG_UBOOT_ARCH}/dts/${CONFIG_UBOOT_DTS}.dtb -j24"
+  eval "CONFIG_RDK_SYS=${CONFIG_RDK_SYS} $cmd2run make ${CONFIG_UBOOT_DEFCONFIG} -j${CONFIG_CPU_NUMBER} V=1 HOSTCC=gcc"
+  eval "CONFIG_RDK_SYS=${CONFIG_RDK_SYS} $cmd2run make EXT_DTB=arch/${CONFIG_UBOOT_ARCH}/dts/${CONFIG_UBOOT_DTS}.dtb -j24 HOSTCC=gcc"
   mv configs/merged_defconfig ${opt_outdir_intermediate}/output_uboot -v
   cd - || exit 1
 }
 #############
