
Upstream-Status: Inappropriate [ astra specific ]

diff --git a/libweston/backend-drm/kms.c b/libweston/backend-drm/kms.c
index ba1c169..4b752e6 100644
--- a/libweston/backend-drm/kms.c
+++ b/libweston/backend-drm/kms.c
@@ -1348,6 +1348,7 @@ drm_pending_state_apply_atomic(struct drm_pending_state *pending_state,
 	struct drm_device *device = pending_state->device;
 	struct drm_backend *b = device->backend;
 	struct drm_output_state *output_state, *tmp;
+	struct drm_plane_state *plane_state;
 	struct drm_plane *plane;
 	drmModeAtomicReq *req = drmModeAtomicAlloc();
 	uint32_t flags, tear_flag = 0;
@@ -1444,13 +1445,15 @@ drm_pending_state_apply_atomic(struct drm_pending_state *pending_state,
 			ret |= crtc_add_prop(req, crtc, WDRM_CRTC_MODE_ID, 0);
 		}
 
-		/* Disable all the planes; planes which are being used will
-		 * override this state in the output-state application. */
-		wl_list_for_each(plane, &device->plane_list, link) {
-			drm_debug(b, "\t\t[atomic] starting with plane %lu disabled\n",
-				  (unsigned long) plane->plane_id);
-			plane_add_prop(req, plane, WDRM_PLANE_CRTC_ID, 0);
-			plane_add_prop(req, plane, WDRM_PLANE_FB_ID, 0);
+		wl_list_for_each(output_state, &pending_state->output_list, link) {
+			if (output_state->output->virtual)
+				continue;
+			wl_list_for_each(plane_state, &output_state->plane_list, link) {
+				plane = plane_state->plane;
+				drm_debug(b,"\t\t[atomic] reset plane : %d\n", plane->plane_id);
+				plane_add_prop(req, plane, WDRM_PLANE_CRTC_ID, 0);
+				plane_add_prop(req, plane, WDRM_PLANE_FB_ID, 0);
+			}
 		}
 
 		flags |= DRM_MODE_ATOMIC_ALLOW_MODESET;
