From bf47f5a2871ddacfc68ade26499cb9890b507d3b Mon Sep 17 00:00:00 2001
From: Hsia-Jun Li <randy.li@synaptics.com>
Date: Wed, 18 Jun 2025 18:26:35 +0800
Subject: [PATCH] disable systemd service file

---
 tee-supplicant/CMakeLists.txt | 9 ---------
 1 file changed, 9 deletions(-)

diff --git a/tee-supplicant/CMakeLists.txt b/tee-supplicant/CMakeLists.txt
index 8df9bef..5a6b2d0 100644
--- a/tee-supplicant/CMakeLists.txt
+++ b/tee-supplicant/CMakeLists.txt
@@ -113,12 +113,3 @@ if(CFG_TEE_SUPP_PLUGINS)
 		PRIVATE dl
 	)
 endif()
-
-################################################################################
-# Install targets
-################################################################################
-install(TARGETS ${PROJECT_NAME} RUNTIME DESTINATION ${CMAKE_INSTALL_SBINDIR})
-configure_file(tee-supplicant@.service.in tee-supplicant@.service @ONLY)
-install(FILES ${CMAKE_BINARY_DIR}/${PROJECT_NAME}/tee-supplicant@.service DESTINATION ${CMAKE_INSTALL_LIBDIR}/systemd/system)
-configure_file(optee-udev.rules.in optee-udev.rules @ONLY)
-install(FILES ${CMAKE_BINARY_DIR}/${PROJECT_NAME}/optee-udev.rules DESTINATION ${CMAKE_INSTALL_SYSCONFDIR}/udev/rules.d)
-- 
2.17.1

