From: Developer Name <developer@synaptics.com>
Subject: glamor: xv: add glsl shaders for OpenGLES 3.0
Upstream-Status: Pending
---
diff --git a/glamor/glamor_transfer.c b/glamor/glamor_transfer.c
index e706e0f..6facbd8 100644
--- a/glamor/glamor_transfer.c
+++ b/glamor/glamor_transfer.c
@@ -36,8 +36,8 @@ glamor_upload_boxes(PixmapPtr pixmap, BoxPtr in_boxes, int in_nbox,
     glamor_screen_private       *glamor_priv = glamor_get_screen_private(screen);
     glamor_pixmap_private       *priv = glamor_get_pixmap_private(pixmap);
     int                         box_index;
-    int                         bytes_per_pixel = pixmap->drawable.bitsPerPixel >> 3;
     const struct glamor_format *f = glamor_format_for_pixmap(pixmap);
+    int                         bytes_per_pixel = PICT_FORMAT_BPP(f->render_format) >> 3;
 
     glamor_make_current(glamor_priv);
 
@@ -136,8 +136,8 @@ glamor_download_boxes(PixmapPtr pixmap, BoxPtr in_boxes, int in_nbox,
     glamor_screen_private *glamor_priv = glamor_get_screen_private(screen);
     glamor_pixmap_private *priv = glamor_get_pixmap_private(pixmap);
     int box_index;
-    int bytes_per_pixel = pixmap->drawable.bitsPerPixel >> 3;
     const struct glamor_format *f = glamor_format_for_pixmap(pixmap);
+    int bytes_per_pixel = PICT_FORMAT_BPP(f->render_format) >> 3;
 
     glamor_make_current(glamor_priv);
 
diff --git a/glamor/glamor_xv.c b/glamor/glamor_xv.c
index 1aca9cc..d016f9d 100644
--- a/glamor/glamor_xv.c
+++ b/glamor/glamor_xv.c
@@ -174,24 +174,22 @@ static const glamor_facet glamor_facet_xv_uyvy = {
                 "in vec2 tcs;\n"
                 ),
     .fs_exec = (
-        "    vec3 uyv;\n"
-        "    vec4 frameOut = texture2D(sampler, tcs.st);\n"
-        "\n"
-        "    vec4 prevPixel = texture2D(sampler, vec2(tcs.s - texelSize.x, tcs.t));\n"
-        "    vec4 nextPixel = texture2D(sampler, vec2(tcs.s + texelSize.x, tcs.t));\n"
-        "\n"
-        "    float delta = 0.50;\n"
-        "\n"
-        "    int even = int(mod(tcs.x / texelSize.x, 2.0));\n"
-        "\n"
-        "    uyv.rgb = float(even)*vec3(frameOut.rg, nextPixel.r) + (1.0-float(even))*vec3(prevPixel.r, frameOut.gr);\n"
-        "\n"
-        "    frameOut.r = uyv.g + 1.403*(uyv.r - delta);\n"
-        "    frameOut.g = uyv.g - 0.714*(uyv.r - delta) - 0.344*(uyv.b - delta);\n"
-        "    frameOut.b = uyv.g + 1.773*(uyv.b - delta);\n"
-        "    frameOut.a = 1.0;\n"
-        "    frag_color = frameOut;\n"
-        ),
+                "        vec4 temp1;\n"
+                "        vec2 xy = texture(sampler, tcs.st).xy;\n"
+                "        vec2 prev_xy = texture(sampler, vec2(tcs.s - texelSize.x, tcs.t)).xy;\n"
+                "        vec2 next_xy = texture(sampler, vec2(tcs.s + texelSize.x, tcs.t)).xy;\n"
+                "\n"
+                "        vec3 sample_yuv;\n"
+                "        int odd = int(mod(tcs.x / texelSize.x, 2.0));\n"
+                "        int even = 1 - odd;\n"
+                "        sample_yuv.yxz = float(even)*vec3(xy, next_xy.x) + float(odd)*vec3(prev_xy.x, xy.yx);\n"
+                "\n"
+                "        temp1.xyz = offsetyco.www * vec3(sample_yuv.x) + offsetyco.xyz;\n"
+                "        temp1.xyz = ucogamma.xyz * vec3(sample_yuv.y) + temp1.xyz;\n"
+                "        temp1.xyz = clamp(vco.xyz * vec3(sample_yuv.z) + temp1.xyz, 0.0, 1.0);\n"
+                "        temp1.w = 1.0;\n"
+                "        frag_color = temp1;\n"
+                ),
 };
 
 static const glamor_facet glamor_facet_xv_uyvy_300es = {
@@ -218,23 +216,21 @@ static const glamor_facet glamor_facet_xv_uyvy_300es = {
         "out vec4 myFragColor;\n"
     ),
     .fs_exec = (
-        "    vec3 uyv;\n"
-        "    vec4 frameOut = texture(sampler, tcs.st);\n"
-        "\n"
-        "    vec4 prevPixel = texture(sampler, vec2(tcs.s - texelSize.x, tcs.t));\n"
-        "    vec4 nextPixel = texture(sampler, vec2(tcs.s + texelSize.x, tcs.t));\n"
-        "\n"
-        "    float delta = 0.50;\n"
-        "\n"
-        "    int even = int(mod(tcs.x / texelSize.x, 2.0));\n"
+        "    vec4 temp1;\n"
+        "    vec2 xy = texture(sampler, tcs.st).xy;\n"
+        "    vec2 prev_xy = texture(sampler, vec2(tcs.s - texelSize.x, tcs.t)).xy;\n"
+        "    vec2 next_xy = texture(sampler, vec2(tcs.s + texelSize.x, tcs.t)).xy;\n"
         "\n"
-        "    uyv.rgb = float(even) * vec3(frameOut.rg, nextPixel.r) + (1.0 - float(even)) * vec3(prevPixel.r, frameOut.gr);\n"
+        "    vec3 sample_yuv;\n"
+        "    int odd = int(mod(tcs.x / texelSize.x, 2.0));\n"
+        "    int even = 1 - odd;\n"
+        "    sample_yuv.yxz = float(even) * vec3(xy, next_xy.x) + float(odd) * vec3(prev_xy.x, xy.yx);\n"
         "\n"
-        "    frameOut.r = uyv.g + 1.403 * (uyv.r - delta);\n"
-        "    frameOut.g = uyv.g - 0.714 * (uyv.r - delta) - 0.344 * (uyv.b - delta);\n"
-        "    frameOut.b = uyv.g + 1.773 * (uyv.b - delta);\n"
-        "    frameOut.a = 1.0;\n"
-        "    myFragColor = frameOut;\n"
+        "    temp1.xyz = offsetyco.www * vec3(sample_yuv.x) + offsetyco.xyz;\n"
+        "    temp1.xyz = ucogamma.xyz * vec3(sample_yuv.y) + temp1.xyz;\n"
+        "    temp1.xyz = clamp(vco.xyz * vec3(sample_yuv.z) + temp1.xyz, 0.0, 1.0);\n"
+        "    temp1.w = 1.0;\n"
+        "    myFragColor = temp1;\n"
     ),
 };
 
@@ -500,9 +496,16 @@ glamor_xv_query_image_attributes(int id,
             offsets[0] = 0;
         size *= *h;
         break;
-    case FOURCC_RGB565:
     case FOURCC_UYVY:
         /* UYVU is single-plane really, all tranformation is processed inside a shader */
+        size = ALIGN(*w, 2) * 2;
+        if (pitches)
+            pitches[0] = size;
+        if (offsets)
+            offsets[0] = 0;
+        size *= *h;
+        break;
+    case FOURCC_RGB565:
         size = *w * 2;
         if (pitches)
             pitches[0] = size;
@@ -908,6 +911,15 @@ glamor_xv_put_image(glamor_port_private *port_priv,
                             buf + s2offset, srcPitch);
         break;
     case FOURCC_UYVY:
+        srcPitch = ALIGN(width, 2) * 2;
+        full_box.x1 = 0;
+        full_box.y1 = 0;
+        full_box.x2 = width;
+        full_box.y2 = height;
+        glamor_upload_boxes(&port_priv->src_pix[0]->drawable, &full_box, 1,
+                            0, 0, 0, 0,
+                            buf, srcPitch);
+        break;
     case FOURCC_RGB565:
         srcPitch = width * 2;
         full_box.x1 = 0;
