diff -ruN git/runtime/torq_hw/hal/SL2610/Makefile git_modified/runtime/torq_hw/hal/SL2610/Makefile
--- git/runtime/torq_hw/hal/SL2610/Makefile	2025-11-06 05:28:07.358209122 +0000
+++ git_modified/runtime/torq_hw/hal/SL2610/Makefile	2025-11-06 05:38:00.818271673 +0000
@@ -1,9 +1,11 @@
 # SPDX-License-Identifier: GPL-2.0-only
 #
+# Build syna_npu.ko as an external module
+obj-m += syna_npu.o
 
-obj-$(CONFIG_SYNAPTICS_NPU) += syna_npu.o
+# Composite object contents (list .o files that come from your .c sources)
 syna_npu-y += src/torq_core_linux.o
 
-HEADER_PATH += -I$(src)/inc/
-ccflags-y += $(HEADER_PATH)
+# Include paths: always reference sources via $(src)
+ccflags-y += -I$(src)/inc
 ccflags-y += -Werror
